Description: Resolve polkit dialog crashes
 errors.ubuntu.com reports many polkit crashes
 over many versions of budgie-desktop.
 This is a patch designed to resolve or reduce
 the number of reports. The reports of
 errors.ubuntu.com will continue to be
 monitored to see if this patch is good
 enough to be forwarded upstream
Author: David Mohammed <fossfreedom@ubuntu.com>
Forwarded: no
Last-Update: 2020-03-18

--- budgie-desktop-10.5.1.orig/src/polkit/polkitdialog.vala
+++ budgie-desktop-10.5.1/src/polkit/polkitdialog.vala
@@ -70,7 +70,7 @@ public class AgentDialog : Gtk.Window
     }
 
     /* Manipulate Vala's pointer logic to prevent a copy */
-    public unowned GLib.Cancellable? cancellable { public set ; public unowned get; }
+    public unowned GLib.Cancellable? cancellable { public set ; public get; }
 
     public string cookie { public get; public set; }
 
@@ -215,7 +215,7 @@ public class AgentDialog : Gtk.Window
     /* This bit is inspired by lxpolkit */
     public void set_from_idents(List<Polkit.Identity?> idents)
     {
-        Gtk.ListBoxRow? active_row = null;
+        //Gtk.ListBoxRow? active_row = null;
 
         Gtk.ListStore? model = new Gtk.ListStore(2, typeof(string), typeof(Polkit.Identity));
         Gtk.TreeIter iter;
@@ -286,6 +286,8 @@ public class Agent : PolkitAgent.Listene
     /* Theme management */
     private Budgie.ThemeManager theme_manager;
 
+    public signal void stop ();
+
     public override async bool initiate_authentication(string action_id, string message, string icon_name,
         Polkit.Details details, string cookie, GLib.List<Polkit.Identity?>? identities, GLib.Cancellable cancellable) throws Polkit.Error
     {
@@ -349,7 +351,8 @@ public class Agent : PolkitAgent.Listene
     private void end_session(bool quit)
     {
         if (quit) {
-            Gtk.main_quit();
+            //Gtk.main_quit();
+            stop();
             return;
         }
         try {
@@ -385,7 +388,10 @@ public static int main(string[] args)
     }
 
     try {
-        PolkitAgent.register_listener(agent, subject, null);
+        PolkitAgent.register_listener(agent, subject, "");
+        agent.stop.connect(()=> {
+            Gtk.main_quit();
+        });
     } catch (Error e) {
         stderr.printf("Unable to register listener: %s", e.message);
         return 1;
